src/Angular.js

```diff
@@ -765,23 +767,37 @@ ES6MapShim.prototype = {
     }
     return (this._lastIndex = (this._keys.indexOf(this._lastKey = key)));
   },
+  _transformKey: function(key) {
+    return isNumberNaN(key) ? nanPlaceholder : key;
+  },
   get: function(key) {
-    var idx = this._idx(key);
+    var idx = this._idx(this._transformKey(key));
     if (idx !== -1) {
       return this._values[idx];
     }
   },
   set: function(key, value) {
+    key = this._transformKey(key);
     var idx = this._idx(key);
     if (idx === -1) {
       idx = this._lastIndex = this._keys.length;
-      this._lastKey = key;
     }
     this._keys[idx] = key;
     this._values[idx] = value;
 
     // Support: IE11
     // Do not `return this` to simulate the partial IE11 implementation
+  },
+  delete: function(key) {
+    var idx = this._idx(this._transformKey(key));
+    if (idx === -1) {
+      return false;
+    }
+    this._keys.splice(idx, 1);
+    this._values.splice(idx, 1);
+    this._lastKey = NaN;
```

Comment:
Only have to reset the `last*` if `idx <= lastIdx`, I think? Not sure if it's worth any extra logic here though...